sources/web/datalab/fileSearch.ts

```diff
@@ -0,0 +1,126 @@
+/// <reference path="../../../third_party/externs/ts/chokidar/chokidar.d.ts" />
+/// <reference path="common.d.ts" />
+
+import http = require('http');
+import url = require('url');
+import fs = require('fs');
+import path = require('path');
+import logging = require('./logging');
+import chokidar = require('chokidar');
+
+let appSettings: common.Settings;
+let fileIndex: string[] = [];
+let indexReady: boolean = false;
+let tooManyFiles: boolean = false;
+const fileCountLimit = 1000000;
+
+// this is matched by the client javascript to display a message that
+// there are more results than shown. Make sure these are in sync
+const clientResultSize = 20; 
+
+/**
+ * Implements the file search/filter request handling.
+ * @param request the incoming search request.
+ * @param response the outgoing search response.
+ */
+function requestHandler(request: http.ServerRequest, response: http.ServerResponse): void {
+  const parsedUrl = url.parse(request.url, true);
+  const pattern = parsedUrl.query['pattern'];  
+  const statusCheck = parsedUrl.query['status'];
+
+  response.writeHead(200, { 'Content-Type': 'application/json' });
+  if (pattern !== undefined) {
+    const results = filter(pattern, fileIndex, clientResultSize);
+    response.write(JSON.stringify(results));
+  } else if (statusCheck !== undefined) {
+    response.write(JSON.stringify({
+      tooManyFiles: tooManyFiles,
+      indexSize: fileIndex.length,
+      indexReady: indexReady
+    }));
+  }
+  response.end();
+}
+
+/**
+ * Builds an index of all files in the content directory to make search faster
+ */
+export function indexFiles(): void {
+  const startTime = process.hrtime();
+  logging.getLogger().info('Started indexing file system');
+  chokidar.watch(appSettings.contentDir + '/', {
+      usePolling: true,
+      interval: 1000,               // we don't need high frequency polling
+      ignored: /(^|[\/\\])\../,     // ignore dot files/dirs
+      ignorePermissionErrors: true, // ignore files with no permissions
+    })
+    .on('add', (addedPath) => {
+      fileIndex.push(addedPath.substr(appSettings.contentDir.length + 1));
+    })
+    .on('unlink', (deletedPath) => {
+      deletedPath = deletedPath.substr(appSettings.contentDir.length + 1);
+      let pos = fileIndex.indexOf(deletedPath);
+      if (pos > -1) {
+        fileIndex.splice(pos, 1);
+      }
+    })
+    .on('ready', () => {
+      indexReady = true;
+      const indexTime = process.hrtime(startTime);
+      logging.getLogger().info('Finished indexing ' + fileIndex.length + ' files in ' + indexTime[0] + ' seconds');
+    })
+    .on('raw', () => {
+      if (!indexReady) {
+        indexReady = true;
+        const indexTime = process.hrtime(startTime);
+        logging.getLogger().info('Finished indexing ' + fileIndex.length + ' files in ' + indexTime[0] + ' seconds');
+        logging.getLogger().error('Indexing threw raw event');
+      }
+    })
+    .on('error', () => {
```

Comment:
It looks like there is no error flag for the client to tell that an error occurred. If the scan gets an error after scanning 10 files out of 1000, will that look the same as a successful scan of 10 total files?